### PR TITLE
add bookmark reached and viseme received event for TTS

### PR DESCRIPTION
**Pull request by yulin-li**

### CODE REVIEW

File: src/common.speech/SynthesisAdapterBase.ts [hunk 1]
```diff
@@ -356,24 +359,59 @@ export class SynthesisAdapterBase implements IDisposable {
...
-                                    this.privSynthesisTurn.currentTextOffset);
-
-                                if (!!this.privSpeechSynthesizer.wordBoundary) {
-                                    try {
-                                        this.privSpeechSynthesizer.wordBoundary(this.privSpeechSynthesizer, ev);
-                                    } catch (error) {
-                                        // Not going to let errors in the event handler
-                                        // trip things up.
+                            switch (metadata.Type) {
+                                case MetadataType.WordBoundary:
```
Comment (glharper) [inline src/common.speech/SynthesisAdapterBase.ts:363]: I like this structure a lot better than testing against string literals!

File: src/common.speech/SynthesisAdapterBase.ts [hunk 2]
```diff
@@ -356,24 +359,59 @@ export class SynthesisAdapterBase implements IDisposable {
...
+                                    this.privSynthesisTurn.onWordBoundaryEvent(metadata.Data.text.Text);
+
+                                    const ev: SpeechSynthesisWordBoundaryEventArgs = new SpeechSynthesisWordBoundaryEventArgs(
```
Comment (glharper) [inline src/common.speech/SynthesisAdapterBase.ts:366]: could "ev," "bev," and "vev" have more descriptive names, like "boundaryEventArgs," "bookmarkEventArgs," and "visemeEventArgs"?

File: src/common.speech/SynthesisAdapterBase.ts [hunk 3]
```diff
@@ -356,24 +359,59 @@ export class SynthesisAdapterBase implements IDisposable {
...
+                                            // Not going to let errors in the event handler
+                                            // trip things up.
+                                        }
+                                    }
+                                    break;
+                                case MetadataType.Viseme:
+
+                                    this.privSynthesisTurn.onVisemeMetadataReceived(metadata);
+
+                                    if (metadata.Data.IsLastAnimation !== false) {
```
Comment (glharper) [inline src/common.speech/SynthesisAdapterBase.ts:399]: if (metadata.Data.IsLastAnimation) {

File: src/common.speech/SynthesisTurn.ts [hunk 4]
```diff
@@ -68,6 +69,12 @@ export class SynthesisTurn {
         return this.privTextOffset;
     }
 
+    public get visemeAnimation(): string {
```
Comment (glharper) [inline src/common.speech/SynthesisTurn.ts:72]: Should setters and getters have side effects, like clearing a private variable here?

File: src/common.speech/SynthesisTurn.ts [hunk 5]
```diff
@@ -202,6 +211,12 @@ export class SynthesisTurn {
         this.updateTextOffset(text);
     }
 
+    public onVisemeMetadataReceived(metadata: ISynthesisMetadata): void {
+        if (metadata.Data.AnimationChunk !== undefined) {
+            this.privPartialVisemeAnimation = this.privPartialVisemeAnimation + metadata.Data.AnimationChunk;
```
Comment (glharper) [inline src/common.speech/SynthesisTurn.ts:216]: Can use += here.

**Review by glharper — COMMENTED** [review]
Can we update an existing test or add a new test to cover the new onAudioStart callback?

File: src/common.speech/SynthesisAdapterBase.ts (hunk 1 above)
Comment (yulin-li) [reply src/common.speech/SynthesisAdapterBase.ts:363]: thanks for the detailed review!

File: src/common.speech/SynthesisAdapterBase.ts (hunk 2 above)
Comment (yulin-li) [reply src/common.speech/SynthesisAdapterBase.ts]: thx, updated them

File: src/common.speech/SynthesisAdapterBase.ts (hunk 3 above)
Comment (yulin-li) [reply src/common.speech/SynthesisAdapterBase.ts]: thx

File: src/common.speech/SynthesisTurn.ts (hunk 4 above)
Comment (yulin-li) [reply src/common.speech/SynthesisTurn.ts]: yes, it's hacky... I updated it to a method instead of a getter

**Comment by yulin-li** [comment]
> Can we update an existing test or add a new test to cover the new onAudioStart callback?
Seems that the onAudioStart() would be triggered only in browser, and could not be tested in current test framework. I manually tested this method and it works.

**Review by glharper — APPROVED** [review]

**Pull request merged**